lib/worker/OperationQueue.js

```diff
@@ -0,0 +1,363 @@
+    this.hasMore = false;
+    // cached queue from the database
+    this.queue = null;
+    // index of the head of the cached queue of operations
+    this.head = 0;
+    // count of possible operations to be added in the chunk
+    this.opCount = 0;
+    // list of operations to be removed
+    this.removedOperations = [];
+    // caches operation removal promise
+    this.removing = null;
+    // caches database empty check promise
+    this.databaseEmpty = null;
+    // caches fetch record promise
+    this.fetchRecords = null;
+    // constants
+    this.MAX_OPERATIONS = _continuityConstants.events.maxOperations;
+    this.MAX_CACHED_QUEUE_SIZE = Math.max(
+      DEFAULT_MAX_CACHED_QUEUE_SIZE,
+      this.MAX_OPERATIONS);
+  }
+
+  /**
+   * Returns whether or not there is another chunk of operations that can
+   * be put into a local regular event. This MUST NOT be called concurrently.
+   *
+   * @return a Promise that resolves to `true` or `false`.
+   */
+  async hasNextChunk() {
+    // if opCount is non-zero then the function has previously been called and
+    // the chunk was not popped from the queue
+    if(this.opCount > 0) {
+      return true;
+    }
+
+    const {ledgerNodeId} = this;
+    // wait for removal of previous chunk before continuing
+    if(this.removing) {
+      await this.removing.catch(e => {
+        logger.error(OPERATIONS_REMOVAL_ERROR, {ledgerNodeId, error: e});
+      });
+    }
+
+    // if the queue is not initialized or the head of the queue is at the end
+    // of the queue, then no next chunk of operations cached yet, so create one
+    // that will be cached when `getNextChunk` is called and its operations are
+    // retrieved
+    if(!(Array.isArray(this.queue) && (this.head + 1) < this.queue.length)) {
+      this.head = 0;
+      this.queue = null;
+
+      const {ledgerNodeId} = this;
+      const projection = {_id: 0, meta: 1, operation: 1};
+
+      if(this.fetchRecords === null) {
+        const {MAX_CACHED_QUEUE_SIZE} = this;
+        this.fetchRecords = this.collection.find({ledgerNodeId}, {projection})
+          .sort({'meta.created': 1}).limit(MAX_CACHED_QUEUE_SIZE).toArray();
+      }
+
+      try {
+        const records = await this.fetchRecords;
+
+        if(records.length === 0) {
+          this.fetchRecords = null;
+          // no new operations
+          return false;
+        }
+
+        this.queue = records;
+      } catch(e) {
+        this.fetchRecords = null;
+        const msg = 'Error fetching operations from queue in database.';
+        logger.error(msg, {ledgerNodeId, error: e});
+        // error fetching records from queue
+        return false;
+      }
+    }
+
+    const {head, opCount} = await this._getHead();
+
+    if(head === null) {
+      this.fetchRecords = null;
+      logger.debug('Unable to find a valid head of operation queue.', {
+        ledgerNodeId
+      });
+      return false;
+    }
```

Comment:
This doesn't quite feel clean -- like we've got some state that is partitioned that doesn't need to be. We initialized `head` back to `0` and `queue` back to `null` but we potentially kept this `fetchRecords` thing around as internal state from some past thing... and awaited it ... and now we're calling `this._getHead()` on that. Can `fetchRecords` just live in here locally? Why does it need to live beyond this function call as instance state?

Sorry, having a hard time following this bit and it seems like it should be simpler -- I may be missing something.

Reply:
This is done to prevent multiple database calls when the function is called back-to-back in rapid succession.

Reply:
But the function is only called by the worker -- and the internal state of the queue should prevent another database call until what was retrieved last time is handled. I therefore don't understand the use case. Is this a remnant of the previous code that  did not have the non-currency assumption (the queue's API will only ever be called serially) built in?

Reply:
> ~~This is done to prevent multiple database calls when the function is called back-to-back in rapid succession~~. 

After reviewing the implementation of `_getHead` , it does have the non-concurrent call assumption. It returns `null` when a valid head could not be found. This means that it found 0 valid operations out of the set of operations in pulled from the database.